lib/plugins/aws/provider.js

```diff
@@ -1704,4 +1744,253 @@ class AwsProvider {
+
+        let isDockerfileAvailable = false;
+        const pathToDockerfile = path.join(imagePath, 'Dockerfile');
+
+        try {
+          const stats = await fs.promises.stat(pathToDockerfile);
+          isDockerfileAvailable = stats.isFile();
+        } catch {
+          // pass and handle after catch block
+        }
+        if (!isDockerfileAvailable) {
+          throw new this.serverless.classes.Error(
+            `Could not access Dockerfile under path: "${pathToDockerfile}"`,
+            'DOCKERFILE_NOT_AVAILABLE_ERROR'
+          );
+        }
+
+        const { repositoryUri, repositoryName } = await this.getOrCreateEcrRepository();
+
+        const localTag = `${repositoryName}:${imageName}`;
+        const remoteTag = `${repositoryUri}:${imageName}`;
+
+        const buildDockerArgs = ['build', '-t', localTag, imagePath];
+        try {
+          const { stdBuffer: buildDockerStdBuffer } = await spawnExt('docker', buildDockerArgs);
+          process.stdout.write(buildDockerStdBuffer);
+        } catch (err) {
+          throw new this.serverless.classes.Error(
+            `Encountered error during executing: docker ${buildDockerArgs.join(
+              ' '
+            )}\nOutput of the command:\n${err.stdBuffer}`,
+            'DOCKER_BUILD_ERROR'
+          );
+        }
+
+        const tagDockerArgs = ['tag', localTag, remoteTag];
+        try {
+          const { stdBuffer: tagDockerStdBuffer } = await spawnExt('docker', tagDockerArgs);
+          process.stdout.write(tagDockerStdBuffer);
+        } catch (err) {
+          throw new this.serverless.classes.Error(
+            `Encountered error during executing: docker ${tagDockerArgs.join(
+              ' '
+            )}\nOutput of the command:\n${err.stdBuffer}`,
+            'DOCKER_TAG_ERROR'
+          );
+        }
+
+        let pushDockerStdBuffer;
+        const pushDockerArgs = ['push', remoteTag];
+
+        try {
+          ({ stdBuffer: pushDockerStdBuffer } = await spawnExt('docker', pushDockerArgs));
+          process.stdout.write(pushDockerStdBuffer);
+        } catch (outerPushErr) {
+          let errToThrow = outerPushErr;
+          if (
+            outerPushErr.stdBuffer &&
+            outerPushErr.stdBuffer.includes('no basic auth credentials')
+          ) {
+            errToThrow = null;
+            await this.dockerLoginToEcr();
+            try {
+              ({ stdBuffer: pushDockerStdBuffer } = await spawnExt('docker', pushDockerArgs));
+              process.stdout.write(pushDockerStdBuffer);
+            } catch (innerPushError) {
+              errToThrow = innerPushError;
+            }
+          }
+          if (errToThrow) {
+            throw new this.serverless.classes.Error(
+              `Encountered error during executing: docker ${pushDockerArgs}\nOutput of the command:\n${errToThrow.stdBuffer}`,
+              'DOCKER_PUSH_ERROR'
+            );
+          }
+        }
```

Comment:
I think it'll be nicer to configure with more _functional_ than _imperative_ approach. Something as:

```javascript
const pushTag = (args, options = {}) => {
  try {
    spawn('push tag', args);
  } catch (error) {
    if (!options.isLoggedIn && notLoggedInError) {
      login();
      return pushTag(args, { isLoggedIn: true })
    }
    throw error;
  }
}

pushTag(args);
```

Reply:
great idea - changed